File: scripts/import_large_osm_xml.ts
/**
 * CLI Script: Stream & Import Large OSM XML (.osm) directly into PostgreSQL
 *
 * Usage:
 *   npx tsx scripts/import_large_osm_xml.ts <path-to-file.osm> [batchSize]
 *
 * Example:
 *   npx tsx scripts/import_large_osm_xml.ts ./vietnam-latest.osm 500
 *
 * Environment variables:
 *   - DATABASE_URL: postgres://user:password@host:5432/dbname
 *   OR
 *   - SQL_HOST, SQL_USER, SQL_PASSWORD, SQL_DB_NAME, SQL_PORT
 */

import fs from 'fs';
import path from 'path';
import sax from 'sax';
import { db } from '../src/db/index.ts';
import { osmPlaces, importLogs } from '../src/db/schema.ts';
import { sql } from 'drizzle-orm';
import { OSMPlace, OSMTags } from '../server/types.ts';
import { computeBoundingBox } from '../server/spatialIndex.ts';
import { getAdminMeta, formatBuildingName } from '../server/pbfParser.ts';

// Bounding box filter for Vietnam (optional safety boundary)
const VN_BBOX = {
  minLon: 102.0,
  maxLon: 110.5,
  minLat: 8.0,
  maxLat: 24.0,
};

function calculateApproxAreaKm2(rings: number[][][]): number {
  if (!rings || rings.length === 0 || !rings[0] || rings[0].length < 3) return 0;
  const ring = rings[0];
  let area = 0;
  for (let i = 0; i < ring.length - 1; i++) {
    const p1 = ring[i];
    const p2 = ring[i + 1];
    area += (p2[0] - p1[0]) * (p2[1] + p1[1]);
  }
  return Math.abs(area / 2) * 111 * 106;
}

async function flushBatch(batch: OSMPlace[], sourceName: string): Promise<number> {
  if (batch.length === 0) return 0;

  const rows = batch.map((place) => ({
    id: place.id,
    osmId: Number(place.osmId),
    osmType: place.osmType,
    name: place.name,
    nameVi: place.nameVi ?? null,
    nameEn: place.nameEn ?? null,
    adminLevel: String(place.adminLevel),
    adminLevelLabel: place.adminLevelLabel,
    priorityRank: place.placeType === 'building' ? 0 : Math.round(Number(place.priorityRank) || 0),
    placeType: place.placeType,
    tags: place.tags || {},
    minLon: place.bbox.minLon,
    minLat: place.bbox.minLat,
    maxLon: place.bbox.maxLon,
    maxLat: place.bbox.maxLat,
    centerLon: place.center[0],
    centerLat: place.center[1],
    geometryType: place.geometryType,
    geometry: place.geometry,
    areaApproxKm2: place.areaApproxKm2 ?? null,
    source: sourceName,
  }));

  await db
    .insert(osmPlaces)
    .values(rows)
    .onConflictDoUpdate({
      target: osmPlaces.id,
      set: {
        name: sql`excluded.name`,
        nameVi: sql`excluded.name_vi`,
        nameEn: sql`excluded.name_en`,
        tags: sql`excluded.tags`,
        minLon: sql`excluded.min_lon`,
        minLat: sql`excluded.min_lat`,
        maxLon: sql`excluded.max_lon`,
        maxLat: sql`excluded.max_lat`,
        centerLon: sql`excluded.center_lon`,
        centerLat: sql`excluded.center_lat`,
        geometryType: sql`excluded.geometry_type`,
        geometry: sql`excluded.geometry`,
        areaApproxKm2: sql`excluded.area_approx_km2`,
        source: sql`excluded.source`,
      },
    });

  const count = batch.length;
  batch.length = 0; // Clear array in place to free memory
  return count;
}

async function run() {
  const args = process.argv.slice(2);
  const filePath = args[0];
  const batchSize = parseInt(args[1] || '500', 10);

  if (!filePath) {
    console.error('❌ Thiếu đường dẫn file XML.');
    console.log('Cách dùng:');
    console.log('  npx tsx scripts/import_large_osm_xml.ts <duong-dan-file.osm> [batchSize]');
    console.log('Ví dụ:');
    console.log('  npx tsx scripts/import_large_osm_xml.ts ./vietnam.osm 500');
    process.exit(1);
  }

  const resolvedPath = path.resolve(filePath);
  if (!fs.existsSync(resolvedPath)) {
    console.error(`❌ Không tìm thấy file: ${resolvedPath}`);
    process.exit(1);
  }

  const stat = fs.statSync(resolvedPath);
  const fileSizeMb = (stat.size / (1024 * 1024)).toFixed(2);
  const fileName = path.basename(resolvedPath);

  console.log('====================================================');
  console.log(`🚀 BẮT ĐẦU ĐỌC VÀ LƯU FILE OSM XML VÀO POSTGRESQL`);
  console.log(`📁 File: ${fileName} (${fileSizeMb} MB)`);
  console.log(`📦 Batch size: ${batchSize} records/insert`);
  console.log('====================================================');

  const startTime = Date.now();
  let totalSaved = 0;
  let batchBuffer: OSMPlace[] = [];

  // Memory map for node coordinates (needed to reconstruct polygons from ways)
  // [lon, lat] packed or stored
  const nodeCoords = new Map<number, [number, number]>();

  interface TempWay {
    id: number;
    refs: number[];
    tags: OSMTags;
  }

  interface TempRelation {
    id: number;
    members: Array<{ type: string; id: number; role: string }>;
    tags: OSMTags;
  }

  const pendingWays = new Map<number, TempWay>();
  const pendingRelations: TempRelation[] = [];

  let currentNode: { id: number; lat: number; lon: number; tags: OSMTags } | null = null;
  let currentWay: { id: number; refs: number[]; tags: OSMTags } | null = null;
  let currentRelation: {
    id: number;
    members: Array<{ type: string; id: number; role: string }>;
    tags: OSMTags;
  } | null = null;

  let nodeCount = 0;
  let wayCount = 0;
  let relationCount = 0;
  let lastProgressTime = Date.now();

  const fileStream = fs.createReadStream(resolvedPath);
  const saxStream = sax.createStream(true, { trim: true });

  fileStream.pipe(saxStream);

  saxStream.on('error', (err: any) => {
    console.error('❌ Lỗi XML SAX Parser:', err.message);
  });

  saxStream.on('opentag', (node: sax.Tag) => {
    const tagName = node.name;
    const attrs = node.attributes as Record<string, string>;

    if (tagName === 'node') {
      const id = parseInt(attrs.id, 10);
      const lat = parseFloat(attrs.lat);
      const lon = parseFloat(attrs.lon);
      if (!isNaN(id) && !isNaN(lat) && !isNaN(lon)) {
        currentNode = { id, lat, lon, tags: {} };
      }
    } else if (tagName === 'way') {
      const id = parseInt(attrs.id, 10);
      if (!isNaN(id)) {
        currentWay = { id, refs: [], tags: {} };
      }
    } else if (tagName === 'relation') {
      const id = parseInt(attrs.id, 10);
      if (!isNaN(id)) {
        currentRelation = { id, members: [], tags: {} };
      }
    } else if (tagName === 'tag') {
      const k = attrs.k;
      const v = attrs.v;
      if (k && v !== undefined) {
        if (currentNode) currentNode.tags[k] = v;
        else if (currentWay) currentWay.tags[k] = v;
        else if (currentRelation) currentRelation.tags[k] = v;
      }
    } else if (tagName === 'nd') {
      if (currentWay && attrs.ref) {
        const ref = parseInt(attrs.ref, 10);
        if (!isNaN(ref)) currentWay.refs.push(ref);
      }
    } else if (tagName === 'member') {
      if (currentRelation && attrs.ref) {
        const ref = parseInt(attrs.ref, 10);
        if (!isNaN(ref)) {
          currentRelation.members.push({
            type: attrs.type || 'way',
            id: ref,
            role: attrs.role || '',
          });
        }
      }
    }
  });

  saxStream.on('closetag', async (tagName: string) => {
    if (tagName === 'node' && currentNode) {
      nodeCount++;
      const { id, lat, lon, tags } = currentNode;

      // Filter in Vietnam or store
      if (lat >= VN_BBOX.minLat && lat <= VN_BBOX.maxLat && lon >= VN_BBOX.minLon && lon <= VN_BBOX.maxLon) {
        nodeCoords.set(id, [lon, lat]);

        const name = tags.name || tags['name:vi'] || tags['name:en'];
        // Check if this node is an interesting POI / Place
        if (
          name &&
          (tags.place ||
            tags.amenity ||
            tags.tourism ||
            tags.historic ||
            tags.leisure ||
            tags.shop ||
            tags.building)
        ) {
          const meta = getAdminMeta(tags);
          batchBuffer.push({
            id: `node/${id}`,
            osmId: id,
            osmType: 'node',
            name,
            nameVi: tags['name:vi'] || name,
            nameEn: tags['name:en'] || '',
            adminLevel: meta.adminLevel,
            adminLevelLabel: meta.adminLevelLabel,
            priorityRank: meta.priorityRank,
            placeType: meta.placeType,
            tags,
            bbox: {
              minLon: lon - 0.001,
              maxLon: lon + 0.001,
              minLat: lat - 0.001,
              maxLat: lat + 0.001,
            },
            center: [lon, lat],
            geometryType: 'Point',
            geometry: { coordinates: [lon, lat] },
            areaApproxKm2: 0.01,
          });

          if (batchBuffer.length >= batchSize) {
            saxStream.pause();
            const saved = await flushBatch(batchBuffer, fileName);
            totalSaved += saved;
            saxStream.resume();
          }
        }
      }
      currentNode = null;
    } else if (tagName === 'way' && currentWay) {
      wayCount++;
      const { id, refs, tags } = currentWay;
      const isBuilding = !!tags.building && tags.building !== 'no';
      const isBoundary = tags.boundary === 'administrative' || tags.admin_level !== undefined;

      if ((isBuilding || isBoundary) && refs.length >= 3) {
        pendingWays.set(id, currentWay);
      }
      currentWay = null;
    } else if (tagName === 'relation' && currentRelation) {
      relationCount++;
      const { tags } = currentRelation;
      if (tags.boundary === 'administrative' || tags.type === 'boundary' || tags.type === 'multipolygon') {
        pendingRelations.push(currentRelation);
      }
      currentRelation = null;
    }

    // Periodic progress print
    const now = Date.now();
    if (now - lastProgressTime > 3000) {
      lastProgressTime = now;
      const memMb = (process.memoryUsage().heapUsed / (1024 * 1024)).toFixed(0);
      process.stdout.write(
        `\r⏳ Tiến trình: ${nodeCount.toLocaleString()} nodes | ${wayCount.toLocaleString()} ways | ${relationCount.toLocaleString()} relations | Đã lưu Postgres: ${totalSaved.toLocaleString()} | RAM: ${memMb}MB`
      );
    }
  });

  await new Promise<void>((resolve, reject) => {
    saxStream.on('end', () => resolve());
    saxStream.on('error', (e) => reject(e));
    fileStream.on('error', (e) => reject(e));
  });

  console.log('\n\n🔄 Đang xử lý Ways (Nhà, Tòa nhà, Ranh giới)...');

  // Process Ways (Buildings & Boundaries)
  for (const [wayId, way] of pendingWays.entries()) {
    const isBuilding = !!way.tags.building && way.tags.building !== 'no';
    const isBoundary = way.tags.boundary === 'administrative';

    const coords: number[][] = [];
    for (const ref of way.refs) {
      const pt = nodeCoords.get(ref);
      if (pt) coords.push(pt);
    }

    if (coords.length < 3) continue;

    // Ensure ring is closed
    const first = coords[0];
    const last = coords[coords.length - 1];
    if (first[0] !== last[0] || first[1] !== last[1]) {
      coords.push([first[0], first[1]]);
    }

    const rings = [coords];
    const bbox = computeBoundingBox(rings);
    const center: [number, number] = [
      (bbox.minLon + bbox.maxLon) / 2,
      (bbox.minLat + bbox.maxLat) / 2,
    ];
    const area = calculateApproxAreaKm2(rings);

    if (isBuilding) {
      const name = formatBuildingName(way.tags, wayId);
      const buildingType = way.tags.building;
      const label =
        buildingType === 'commercial' || buildingType === 'office'
          ? 'Tòa nhà Văn phòng / Thương mại'
          : buildingType === 'apartments'
          ? 'Chung cư / Khu căn hộ'
          : buildingType === 'hospital'
          ? 'Bệnh viện / Cơ sở Y tế'
          : buildingType === 'school' || buildingType === 'university'
          ? 'Trường học / Cơ sở Giáo dục'
          : 'Công trình / Tòa nhà';

      batchBuffer.push({
        id: `way/${wayId}`,
        osmId: wayId,
        osmType: 'way',
        name,
        nameVi: way.tags['name:vi'] || name,
        nameEn: way.tags['name:en'] || '',
        adminLevel: 'building',
        adminLevelLabel: label,
        priorityRank: 0.5,
        placeType: 'building',
        tags: way.tags,
        bbox,
        center,
        geometryType: 'Polygon',
        geometry: { coordinates: rings },
        areaApproxKm2: area,
      });
    } else if (isBoundary) {
      const name = way.tags.name || way.tags['name:vi'] || `Ranh giới #${wayId}`;
      const meta = getAdminMeta(way.tags);
      batchBuffer.push({
        id: `way/${wayId}`,
        osmId: wayId,
        osmType: 'way',
        name,
        nameVi: way.tags['name:vi'] || name,
        nameEn: way.tags['name:en'] || '',
        adminLevel: meta.adminLevel,
        adminLevelLabel: meta.adminLevelLabel,
        priorityRank: meta.priorityRank,
        placeType: meta.placeType,
        tags: way.tags,
        bbox,
        center,
        geometryType: 'Polygon',
        geometry: { coordinates: rings },
        areaApproxKm2: area,
      });
    }

    if (batchBuffer.length >= batchSize) {
      const saved = await flushBatch(batchBuffer, fileName);
      totalSaved += saved;
      process.stdout.write(`\r💾 Đang lưu Ways vào Postgres: ${totalSaved.toLocaleString()} địa điểm...`);
    }
  }

  console.log('\n🔄 Đang xử lý Relations (Quan hệ Hành chính Cấp 2, 4, 6, 8)...');

  // Process Relations
  for (const rel of pendingRelations) {
    const meta = getAdminMeta(rel.tags);
    const name = rel.tags.name || rel.tags['name:vi'] || rel.tags['name:en'];
    if (!name) continue;

    // Collect outer ways
    const outerWays: number[][][] = [];
    for (const member of rel.members) {
      if (member.type === 'way') {
        const way = pendingWays.get(member.id);
        if (way && way.refs.length >= 2) {
          const wayCoords: number[][] = [];
          for (const ref of way.refs) {
            const pt = nodeCoords.get(ref);
            if (pt) wayCoords.push(pt);
          }
          if (wayCoords.length >= 2) {
            outerWays.push(wayCoords);
          }
        }
      }
    }

    if (outerWays.length === 0) continue;

    const flattened = outerWays.flat();
    if (flattened.length < 3) continue;

    const ringCoords = [flattened];
    // Close ring if not closed
    const firstPt = flattened[0];
    const lastPt = flattened[flattened.length - 1];
    if (firstPt[0] !== lastPt[0] || firstPt[1] !== lastPt[1]) {
      flattened.push([firstPt[0], firstPt[1]]);
    }

    const bbox = computeBoundingBox(ringCoords);
    const center: [number, number] = [
      (bbox.minLon + bbox.maxLon) / 2,
      (bbox.minLat + bbox.maxLat) / 2,
    ];
    const area = calculateApproxAreaKm2(ringCoords);

    batchBuffer.push({
      id: `rel/${rel.id}`,
      osmId: rel.id,
      osmType: 'relation',
      name,
      nameVi: rel.tags['name:vi'] || name,
      nameEn: rel.tags['name:en'] || '',
      adminLevel: meta.adminLevel,
      adminLevelLabel: meta.adminLevelLabel,
      priorityRank: meta.priorityRank,
      placeType: meta.placeType,
      tags: rel.tags,
      bbox,
      center,
      geometryType: 'Polygon',
      geometry: { coordinates: ringCoords },
      areaApproxKm2: area,
    });

    if (batchBuffer.length >= batchSize) {
      const saved = await flushBatch(batchBuffer, fileName);
      totalSaved += saved;
      process.stdout.write(`\r💾 Đang lưu Relations vào Postgres: ${totalSaved.toLocaleString()} địa điểm...`);
    }
  }

  // Flush remaining buffer
  if (batchBuffer.length > 0) {
    const saved = await flushBatch(batchBuffer, fileName);
    totalSaved += saved;
  }

  // Record audit log
  await db.insert(importLogs).values({
    filename: fileName,
    placesCount: totalSaved,
    fileSize: stat.size,
    status: 'success',
  });

  const durationSec = ((Date.now() - startTime) / 1000).toFixed(1);
  console.log('\n\n====================================================');
  console.log(`✅ HOÀN TẤT IMPORT FILE OSM XML VÀO POSTGRESQL!`);
  console.log(`⏱️ Thời gian thực thi: ${durationSec}s`);
  console.log(`📍 Tổng số địa điểm lưu vào database: ${totalSaved.toLocaleString()}`);
  console.log(`🏠 Cơ sở dữ liệu Postgres: Bảng osm_places đã được cập nhật thành công.`);
  console.log('====================================================');
  process.exit(0);
}

run().catch((err) => {
  console.error('\n❌ Có lỗi xảy ra trong quá trình import:', err);
  process.exit(1);
});

File: scripts/import_large_osm_xml.py
#!/usr/bin/env python3
"""
Python Script: Import large OSM XML (.osm) directly into PostgreSQL database.
Uses streaming iterparse (O(1) memory per tag) and psycopg2 execute_values for high throughput.

Requirements:
    pip install psycopg2-binary

Usage:
    export DATABASE_URL="postgresql://user:password@host:port/dbname"
    python3 scripts/import_large_osm_xml.py /path/to/vietnam.osm [batch_size]

Example:
    python3 scripts/import_large_osm_xml.py ./vietnam-latest.osm 500
"""

import sys
import os
import json
import time
import math
import xml.etree.cElementTree as ET
import psycopg2
from psycopg2.extras import execute_values

# Vietnam approximate Bounding Box filter
VN_BBOX = {
    'min_lon': 102.0,
    'max_lon': 110.5,
    'min_lat': 8.0,
    'max_lat': 24.0,
}

def get_admin_meta(tags):
    admin_level_raw = tags.get('admin_level')
    boundary = tags.get('boundary')
    place = tags.get('place')
    building = tags.get('building')

    if admin_level_raw:
        try:
            lvl = int(admin_level_raw)
            if lvl == 2:
                return 2, 'Quốc gia', 7, 'country'
            elif lvl == 3:
                return 3, 'Vùng địa lý', 6, 'region'
            elif lvl == 4:
                return 4, 'Tỉnh / Thành phố trực thuộc TW', 5, 'province'
            elif lvl == 6:
                return 6, 'Quận / Huyện / Thị xã', 4, 'district'
            elif lvl == 8:
                return 8, 'Phường / Xã / Thị trấn', 3, 'ward'
            else:
                return lvl, f'Cấp hành chính {lvl}', 2, 'subdistrict'
        except ValueError:
            pass

    if building and building != 'no':
        return 'building', 'Tòa nhà / Công trình', 0, 'building'

    if place == 'country':
        return 2, 'Quốc gia', 7, 'country'
    elif place in ('state', 'province'):
        return 4, 'Tỉnh / Thành phố', 5, 'province'
    elif place in ('city', 'county', 'district'):
        return 6, 'Quận / Huyện / Thành phố', 4, 'district'
    elif place in ('town', 'village', 'suburb', 'quarter'):
        return 8, 'Phường / Xã / Thị trấn', 3, 'ward'

    return 'poi', 'Địa điểm / Tiện ích (POI)', 0, 'poi'

def compute_bounding_box(coords):
    min_lon = min(p[0] for p in coords)
    max_lon = max(p[0] for p in coords)
    min_lat = min(p[1] for p in coords)
    max_lat = max(p[1] for p in coords)
    return min_lon, min_lat, max_lon, max_lat

def calculate_approx_area_km2(coords):
    if len(coords) < 3:
        return 0.0
    area = 0.0
    for i in range(len(coords) - 1):
        p1 = coords[i]
        p2 = coords[i + 1]
        area += (p2[0] - p1[0]) * (p2[1] + p1[1])
    return abs(area / 2.0) * 111.0 * 106.0

def flush_batch(cursor, batch, source_name):
    if not batch:
        return 0

    query = """
    INSERT INTO osm_places (
        id, osm_id, osm_type, name, name_vi, name_en,
        admin_level, admin_level_label, priority_rank, place_type,
        tags, min_lon, min_lat, max_lon, max_lat,
        center_lon, center_lat, geometry_type, geometry,
        area_approx_km2, source
    ) VALUES %s
    ON CONFLICT (id) DO UPDATE SET
        name = EXCLUDED.name,
        name_vi = EXCLUDED.name_vi,
        name_en = EXCLUDED.name_en,
        tags = EXCLUDED.tags,
        min_lon = EXCLUDED.min_lon,
        min_lat = EXCLUDED.min_lat,
        max_lon = EXCLUDED.max_lon,
        max_lat = EXCLUDED.max_lat,
        center_lon = EXCLUDED.center_lon,
        center_lat = EXCLUDED.center_lat,
        geometry_type = EXCLUDED.geometry_type,
        geometry = EXCLUDED.geometry,
        area_approx_km2 = EXCLUDED.area_approx_km2,
        source = EXCLUDED.source;
    """

    records = []
    for item in batch:
        records.append((
            item['id'],
            item['osm_id'],
            item['osm_type'],
            item['name'],
            item.get('name_vi'),
            item.get('name_en'),
            str(item['admin_level']),
            item['admin_level_label'],
            item['priority_rank'],
            item['place_type'],
            json.dumps(item['tags']),
            item['min_lon'],
            item['min_lat'],
            item['max_lon'],
            item['max_lat'],
            item['center_lon'],
            item['center_lat'],
            item['geometry_type'],
            json.dumps(item['geometry']),
            item.get('area_approx_km2'),
            source_name
        ))

    execute_values(cursor, query, records)
    count = len(batch)
    batch.clear()
    return count

def main():
    if len(sys.argv) < 2:
        print("❌ Thiếu đường dẫn file XML.")
        print("Cách dùng:")
        print("    python3 scripts/import_large_osm_xml.py <duong-dan-file.osm> [batch_size]")
        sys.exit(1)

    xml_path = os.path.abspath(sys.argv[1])
    batch_size = int(sys.argv[2]) if len(sys.argv) > 2 else 500

    if not os.path.exists(xml_path):
        print(f"❌ Không tìm thấy file: {xml_path}")
        sys.exit(1)

    file_size_bytes = os.path.getsize(xml_path)
    file_size_mb = file_size_bytes / (1024 * 1024)
    file_name = os.path.basename(xml_path)

    # Database connection
    db_url = os.getenv("DATABASE_URL")
    if not db_url:
        host = os.getenv("SQL_HOST") or os.getenv("PGHOST") or "localhost"
        user = os.getenv("SQL_USER") or os.getenv("PGUSER") or "postgres"
        password = os.getenv("SQL_PASSWORD") or os.getenv("PGPASSWORD") or ""
        dbname = os.getenv("SQL_DB_NAME") or os.getenv("PGDATABASE") or "osm_vietnam"
        port = os.getenv("SQL_PORT") or os.getenv("PGPORT") or "5432"
        db_url = f"postgresql://{user}:{password}@{host}:{port}/{dbname}"

    print("====================================================")
    print("🚀 PYTHON STREAMING OSM XML IMPORTER")
    print(f"📁 File: {file_name} ({file_size_mb:.2f} MB)")
    print(f"📦 Batch Size: {batch_size}")
    print("====================================================")

    conn = psycopg2.connect(db_url)
    conn.autocommit = True
    cur = conn.cursor()

    start_time = time.time()
    total_saved = 0
    batch_buffer = []

    # Map of nodeId -> (lon, lat)
    node_coords = {}
    pending_ways = {} # wayId -> {'refs': [...], 'tags': {...}}
    pending_relations = []

    node_count = 0
    way_count = 0
    rel_count = 0
    last_print = time.time()

    # Pass 1: Streaming iterparse
    context = ET.iterparse(xml_path, events=('end',))

    for event, elem in context:
        tag = elem.tag

        if tag == 'node':
            node_count += 1
            node_id = int(elem.attrib['id'])
            lat = float(elem.attrib['lat'])
            lon = float(elem.attrib['lon'])

            tags = {}
            for child in elem:
                if child.tag == 'tag':
                    tags[child.attrib['k']] = child.attrib['v']

            if (VN_BBOX['min_lat'] <= lat <= VN_BBOX['max_lat'] and
                VN_BBOX['min_lon'] <= lon <= VN_BBOX['max_lon']):
                node_coords[node_id] = (lon, lat)

                name = tags.get('name') or tags.get('name:vi') or tags.get('name:en')
                if name and (tags.get('place') or tags.get('amenity') or tags.get('tourism') or
                             tags.get('historic') or tags.get('leisure') or tags.get('shop')):
                    admin_lvl, admin_lbl, rank, p_type = get_admin_meta(tags)
                    batch_buffer.append({
                        'id': f'node/{node_id}',
                        'osm_id': node_id,
                        'osm_type': 'node',
                        'name': name,
                        'name_vi': tags.get('name:vi', name),
                        'name_en': tags.get('name:en', ''),
                        'admin_level': admin_lvl,
                        'admin_level_label': admin_lbl,
                        'priority_rank': rank,
                        'place_type': p_type,
                        'tags': tags,
                        'min_lon': lon - 0.001,
                        'max_lon': lon + 0.001,
                        'min_lat': lat - 0.001,
                        'max_lat': lat + 0.001,
                        'center_lon': lon,
                        'center_lat': lat,
                        'geometry_type': 'Point',
                        'geometry': {'coordinates': [lon, lat]},
                        'area_approx_km2': 0.01
                    })

                    if len(batch_buffer) >= batch_size:
                        total_saved += flush_batch(cur, batch_buffer, file_name)

            elem.clear()

        elif tag == 'way':
            way_count += 1
            way_id = int(elem.attrib['id'])
            refs = []
            tags = {}
            for child in elem:
                if child.tag == 'nd':
                    refs.append(int(child.attrib['ref']))
                elif child.tag == 'tag':
                    tags[child.attrib['k']] = child.attrib['v']

            is_bldg = bool(tags.get('building') and tags.get('building') != 'no')
            is_bndry = bool(tags.get('boundary') == 'administrative' or tags.get('admin_level'))

            if (is_bldg or is_bndry) and len(refs) >= 3:
                pending_ways[way_id] = {'refs': refs, 'tags': tags}

            elem.clear()

        elif tag == 'relation':
            rel_count += 1
            rel_id = int(elem.attrib['id'])
            members = []
            tags = {}
            for child in elem:
                if child.tag == 'member':
                    members.append({
                        'type': child.attrib.get('type', 'way'),
                        'id': int(child.attrib['ref']),
                        'role': child.attrib.get('role', '')
                    })
                elif child.tag == 'tag':
                    tags[child.attrib['k']] = child.attrib['v']

            if tags.get('boundary') == 'administrative' or tags.get('type') in ('boundary', 'multipolygon'):
                pending_relations.append({'id': rel_id, 'members': members, 'tags': tags})

            elem.clear()

        # Periodic status
        now = time.time()
        if now - last_print > 3:
            last_print = now
            print(f"\r⏳ Tiến trình: {node_count:,} nodes | {way_count:,} ways | {rel_count:,} relations | Đã lưu: {total_saved:,}", end='', flush=True)

    print(f"\n🔄 Đang xử lý {len(pending_ways):,} ways...")

    # Process Ways
    for way_id, way in pending_ways.items():
        refs = way['refs']
        tags = way['tags']
        is_bldg = bool(tags.get('building') and tags.get('building') != 'no')
        is_bndry = bool(tags.get('boundary') == 'administrative')

        coords = [node_coords[ref] for ref in refs if ref in node_coords]
        if len(coords) < 3:
            continue

        if coords[0] != coords[-1]:
            coords.append(coords[0])

        min_lon, min_lat, max_lon, max_lat = compute_bounding_box(coords)
        center_lon = (min_lon + max_lon) / 2.0
        center_lat = (min_lat + max_lat) / 2.0
        area = calculate_approx_area_km2(coords)

        admin_lvl, admin_lbl, rank, p_type = get_admin_meta(tags)
        name = tags.get('name') or tags.get('name:vi') or (f"Công trình #{way_id}" if is_bldg else f"Ranh giới #{way_id}")

        batch_buffer.append({
            'id': f'way/{way_id}',
            'osm_id': way_id,
            'osm_type': 'way',
            'name': name,
            'name_vi': tags.get('name:vi', name),
            'name_en': tags.get('name:en', ''),
            'admin_level': admin_lvl,
            'admin_level_label': admin_lbl,
            'priority_rank': 0 if is_bldg else rank,
            'place_type': 'building' if is_bldg else p_type,
            'tags': tags,
            'min_lon': min_lon,
            'max_lon': max_lon,
            'min_lat': min_lat,
            'max_lat': max_lat,
            'center_lon': center_lon,
            'center_lat': center_lat,
            'geometry_type': 'Polygon',
            'geometry': {'coordinates': [coords]},
            'area_approx_km2': area
        })

        if len(batch_buffer) >= batch_size:
            total_saved += flush_batch(cur, batch_buffer, file_name)

    print(f"\n🔄 Đang xử lý {len(pending_relations):,} relations...")

    # Process Relations
    for rel in pending_relations:
        rel_id = rel['id']
        tags = rel['tags']
        name = tags.get('name') or tags.get('name:vi') or tags.get('name:en')
        if not name:
            continue

        outer_coords = []
        for m in rel['members']:
            if m['type'] == 'way' and m['id'] in pending_ways:
                w_refs = pending_ways[m['id']]['refs']
                w_pts = [node_coords[r] for r in w_refs if r in node_coords]
                if len(w_pts) >= 2:
                    outer_coords.extend(w_pts)

        if len(outer_coords) < 3:
            continue

        if outer_coords[0] != outer_coords[-1]:
            outer_coords.append(outer_coords[0])

        min_lon, min_lat, max_lon, max_lat = compute_bounding_box(outer_coords)
        center_lon = (min_lon + max_lon) / 2.0
        center_lat = (min_lat + max_lat) / 2.0
        area = calculate_approx_area_km2(outer_coords)

        admin_lvl, admin_lbl, rank, p_type = get_admin_meta(tags)

        batch_buffer.append({
            'id': f'rel/{rel_id}',
            'osm_id': rel_id,
            'osm_type': 'relation',
            'name': name,
            'name_vi': tags.get('name:vi', name),
            'name_en': tags.get('name:en', ''),
            'admin_level': admin_lvl,
            'admin_level_label': admin_lbl,
            'priority_rank': rank,
            'place_type': p_type,
            'tags': tags,
            'min_lon': min_lon,
            'max_lon': max_lon,
            'min_lat': min_lat,
            'max_lat': max_lat,
            'center_lon': center_lon,
            'center_lat': center_lat,
            'geometry_type': 'Polygon',
            'geometry': {'coordinates': [outer_coords]},
            'area_approx_km2': area
        })

        if len(batch_buffer) >= batch_size:
            total_saved += flush_batch(cur, batch_buffer, file_name)

    # Flush final
    if batch_buffer:
        total_saved += flush_batch(cur, batch_buffer, file_name)

    # Audit log
    cur.execute(
        "INSERT INTO import_logs (filename, places_count, file_size, status) VALUES (%s, %s, %s, %s)",
        (file_name, total_saved, file_size_bytes, 'success')
    )

    cur.close()
    conn.close()

    duration = time.time() - start_time
    print(f"\n====================================================")
    print(f"✅ THÀNH CÔNG: Đã lưu {total_saved:,} địa điểm vào PostgreSQL trong {duration:.1f}s!")
    print(f"====================================================")

if __name__ == '__main__':
    main()
